builtin/common/serialize.lua

```diff
@@ -1,205 +1,232 @@
+			return write("true")
+		end
+		if value == false then
+			return write("false")
+		end
+		local type_ = type(value)
+		if type_ == "number" then
+			return write(string_format("%.17g", value))
+		end
+		-- Reference types: table, function and string
+		local ref = references[value]
+		if ref then
+			return write(ref)
+		end
+		if type_ == "string" then
+			return write(quote(value))
+		end
+		if type_ == "function" then
+			return write(dump_func(value))
+		end
+		if type_ == "table" then
+			write("{")
+			-- First write list keys:
+			-- Don't use the table length #value here as it may horribly fail
+			-- for tables which use large integers as keys in the hash part;
+			-- stop at the first "hole" (nil value) instead by using ipairs
+			local len = 0
+			for i, v in ipairs(value) do
+				dump(v)
+				write(";")
+				len = i
 			end
-			for k, v in pairs(x) do
-				if (not np or not np[k]) and
-						not idx_dumped[k] then
-					vals[#vals + 1] = "["..dump_or_ref_val(k).."] = "
-						..dump_or_ref_val(v)
+			-- Now write map keys ([key] = value)
+			for k, v in next, value do
+				-- Check whether this is a non-list key (hash key)
+				if type(k) ~= "number" or k % 1 ~= 0 or k < 1 or k > len then
+					if use_short_key(k) then
+						write(k)
+					else
+						write("[")
+						dump(k)
+						write("]")
+					end
+					write("=")
+					dump(v)
+					write(";")
 				end
 			end
-			return "{"..table.concat(vals, ", ").."}"
-		else
-			error("Can't serialize data of type "..tp)
+			write("}")
+			return
 		end
 	end
-
-	local function dump_nest_points()
-		for parent, vals in pairs(nest_points) do
-			for k, v in pairs(vals) do
-				local_defs[#local_defs + 1] = dump_or_ref_val(parent)
-					.."["..dump_or_ref_val(k).."] = "
-					..dump_or_ref_val(v)
+	for table, ref in pairs(to_fill) do
+		for k, v in pairs(table) do
+			write(ref)
+			if use_short_key(k) then
+				write(".")
+				write(k)
+			else
+				write("[")
+				dump(k)
+				write("]")
 			end
+			write("=")
+			dump(v)
+			write(";")
 		end
 	end
+	write("return ")
+	dump(value)
+end
 
-	mark_multiple_occurences(x)
-	local top_level = dump_or_ref_val(x)
-	dump_nest_points()
-
-	if next(local_defs) then
-		return "local _ = {}\n"
-			..table.concat(local_defs, "\n")
-			.."\nreturn "..top_level
-	else
-		return "return "..top_level
-	end
+function core.serialize(value)
+	local rope = {}
+	serialize(value, function(text)
+		table_insert(rope, text)
+	end)
+	return table_concat(rope)
 end
 
--- Deserialization
+local function dummy_func() end
 
-local function safe_loadstring(...)
-	local func, err = loadstring(...)
-	if func then
-		setfenv(func, {})
-		return func
+local function deserialize(func, safe)
+	-- math.huge is serialized to inf, 0/0 is serialized to -nan
+	local env = {inf = math_huge, nan = 0/0}
+	if safe then
+		env.loadstring = dummy_func
+	else
+		env.loadstring = function(str, ...)
+			if str:byte(1) == 0x1B then
+				return nil, "Bytecode prohibited"
+			end
```

Comment:
handled by mod security and should be removed

Reply:
done